src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBoxViewModelFactory.cs

```diff
@@ -0,0 +1,129 @@
+            _textEditorFactoryService = textEditorFactoryService;
+            _serviceProvider = (OLE.Interop.IServiceProvider)services.GetService(typeof(OLE.Interop.IServiceProvider));
+            _projectionBufferFactoryService = projectionBufferFactoryService;
+        }
+
+        public async Task<IntellisenseTextBoxViewModel[]> CreateIntellisenseTextBoxViewModelsAsync(
+          Document document,
+          IContentType contentType,
+          string contentString,
+          Func<IProjectionSnapshot, ITrackingSpan[]> createSpansMethod,
+          string[][] rolesCollections, CancellationToken cancellationToken)
+        {
+            var vsTextLines = _editorAdaptersFactoryService.CreateVsTextBufferAdapter(_serviceProvider, contentType) as IVsTextLines;
+            vsTextLines.InitializeContent(contentString, contentString.Length);
+
+            var originalContextBuffer = _editorAdaptersFactoryService.GetDataBuffer(vsTextLines);
+            // Get the workspace, and from there, the solution and document containing this buffer.
+            // If there's an ExternalSource, we won't get a document. Give up in that case.
+            var solution = document.Project.Solution;
+
+            // Wrap the original ContextBuffer in a projection buffer that we can make read-only
+            var contextBuffer = _projectionBufferFactoryService.CreateProjectionBuffer(null,
+                new object[] { originalContextBuffer.CurrentSnapshot.CreateFullTrackingSpan(SpanTrackingMode.EdgeInclusive) }, ProjectionBufferOptions.None, contentType);
+
+            // Make projection readonly so we can't edit it by mistake.
+            using (var regionEdit = contextBuffer.CreateReadOnlyRegionEdit())
+            {
+                regionEdit.CreateReadOnlyRegion(new Span(0, contextBuffer.CurrentSnapshot.Length), SpanTrackingMode.EdgeInclusive, EdgeInsertionMode.Deny);
+                regionEdit.Apply();
+            }
+
+            // Put it all into a projection buffer
+            var projectionBuffer = _projectionBufferFactoryService.CreateProjectionBuffer(null,
+                createSpansMethod(contextBuffer.CurrentSnapshot),
+                ProjectionBufferOptions.None, contentType);
+
+            // Fork the solution using this new primary buffer for the document and all of its linked documents.
+            var forkedSolution = solution.WithDocumentText(document.Id, projectionBuffer.CurrentSnapshot.AsText(), PreservationMode.PreserveIdentity);
+            foreach (var link in document.GetLinkedDocumentIds())
+            {
+                forkedSolution = forkedSolution.WithDocumentText(link, projectionBuffer.CurrentSnapshot.AsText(), PreservationMode.PreserveIdentity);
+            }
+
+            // Put it into a new workspace, and open it and its related documents
+            // with the projection buffer as the text.
+            var workspace = new IntellisenseTextBoxWorkspace(forkedSolution, document.Project);
```

Comment:
Is anybody disposing this Workspace once we're all done? It's not _strictly_ required, but we have some test to ensure we don't leak all of this?